### PR TITLE
[User] Use typehints introduced by PHP7+

### PR DESCRIPTION
**Pull request by pamil**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | yes |
| Related tickets | part of #8425 |
| License         | MIT |

### CODE REVIEW

File: src/Sylius/Bundle/UserBundle/EventListener/UserDeleteListener.php [hunk 1]
```diff
@@ -25,17 +25,17 @@
  * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
  * @author Michał Marcinkowski <michal.marcinkowski@lakion.com>
  */
-class UserDeleteListener
+final class UserDeleteListener
```
Comment (Zales0123) [inline src/Sylius/Bundle/UserBundle/EventListener/UserDeleteListener.php]: It's worth to mention it in `UPGRADE.md` file.

File: src/Sylius/Bundle/UserBundle/EventListener/UserLastLoginSubscriber.php [hunk 2]
```diff
@@ -21,12 +21,12 @@
 use Symfony\Component\Security\Http\Event\InteractiveLoginEvent;
 use Symfony\Component\Security\Http\SecurityEvents;
 
-class UserLastLoginSubscriber implements EventSubscriberInterface
+final class UserLastLoginSubscriber implements EventSubscriberInterface
```
Comment (Zales0123) [inline src/Sylius/Bundle/UserBundle/EventListener/UserLastLoginSubscriber.php]: And here.

Comment (Zales0123) [reply src/Sylius/Bundle/UserBundle/EventListener/UserLastLoginSubscriber.php]: And everything else.

File: src/Sylius/Component/Core/Model/CustomerInterface.php [hunk 3]
```diff
@@ -61,16 +61,6 @@ public function hasAddress(AddressInterface $address);
...
 
     /**
-     * @return BaseUserInterface
-     */
-    public function getUser();
-
-    /**
-     * @param BaseUserInterface|null $user
-     */
-    public function setUser(BaseUserInterface $user = null);
```
Comment (Zales0123) [inline src/Sylius/Component/Core/Model/CustomerInterface.php]: We could mention deleting these methods in `UPGRADE` file.

File: src/Sylius/Bundle/CoreBundle/Validator/Constraints/UniqueReviewerEmailValidator.php [hunk 4]
```diff
@@ -64,12 +64,16 @@ public function __construct(
...
     {
-        /* @var ReviewerInterface $customer */
+        /* @var ReviewerInterface|null $customer */
         $customer = $review->getAuthor();
 
         $token = $this->tokenStorage->getToken();
-        if ($this->checkIfUserIsAuthenticated($token)) {
-            if (null !== $customer && $token->getUser()->getCustomer()->getEmail() === $customer->getEmail()) {
+        if (null !== $customer) {
+            if ($customer->getEmail() === null) {
```
[the remark this answers is not in the document]
Comment (pamil) [reply src/Sylius/Bundle/CoreBundle/Validator/Constraints/UniqueReviewerEmailValidator.php]: Please, no yoda :( There's no chance I'd miss two equal signs :D

[the remark this answers is not in the document]
Comment (pamil) [reply src/Sylius/Bundle/CoreBundle/Validator/Constraints/UniqueReviewerEmailValidator.php]: But yeah, it'll look more concise that way.

[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Sylius/Bundle/CoreBundle/Validator/Constraints/UniqueReviewerEmailValidator.php]: You have yoda in different files 🍏  I just want to keep consistency

File: src/Sylius/Component/Core/Model/CustomerInterface.php (hunk 3 above)
Comment (pamil) [reply src/Sylius/Component/Core/Model/CustomerInterface.php]: It's the part of `UserAwareInterface`.

File: src/Sylius/Bundle/UserBundle/EventListener/UserLastLoginSubscriber.php (hunk 2 above)
Comment (pamil) [reply src/Sylius/Bundle/UserBundle/EventListener/UserLastLoginSubscriber.php]: Done.

File: src/Sylius/Bundle/UserBundle/EventListener/UserDeleteListener.php (hunk 1 above)
Comment (pamil) [reply src/Sylius/Bundle/UserBundle/EventListener/UserDeleteListener.php]: Done.

**Comment by Zales0123** [comment]
Pretty good job Kamil, thank you! 🏒

**Pull request merged**